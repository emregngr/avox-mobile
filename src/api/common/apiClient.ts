import AsyncStorage from '@react-native-async-storage/async-storage'
import type { AxiosError, AxiosResponse, InternalAxiosRequestConfig } from 'axios'
import axios from 'axios'
import { router } from 'expo-router'

import { ENUMS } from '@/enums'
import type { ApiError, ApiRequestConfig, ApiResponse } from '@/types/common/api'

const axiosInstance = axios.create({
  baseURL: process.env.EXPO_PUBLIC_API_URL ?? '',
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
  timeout: 60000,
})

axiosInstance.interceptors.request.use(
  async (config: InternalAxiosRequestConfig): Promise<InternalAxiosRequestConfig> => {
    const token = await AsyncStorage.getItem(ENUMS.API_TOKEN)

    if (token && config.headers) {
      config.headers.Authorization = `Bearer ${token}`
    }

    return config
  },
  error => {
    console.warn('HTTP :: Request Error', error)
    return Promise.reject(error)
  },
)

axiosInstance.interceptors.response.use(
  (response: AxiosResponse) => {
    if (response?.status === 401) {
      router.navigate('/token-expire')
      return Promise.reject(new Error())
    }
    return response
  },

  async (error: AxiosError) => {
    const originalRequest = error.config as ApiRequestConfig

    const apiError: ApiError = {
      errors: {},
      message: 'An error occurred',
      status: error.response?.status ?? 500,
    }

    if (error.response?.data) {
      const errorData = error.response.data as any

      if (errorData.message) {
        apiError.message = errorData.message
      }

      if (errorData.errors) {
        apiError.errors = errorData.errors
      }
    }

    if (!originalRequest?.skipErrorHandling) {
      console.warn('Error Status', apiError.status)
      console.warn('Error Message', apiError.message)
      console.warn('Errors', apiError.errors)
    }

    if (error?.message === 'Request failed with status code 401') {
      router.navigate('/token-expire')
      return Promise.reject(error)
    }

    return Promise.reject(apiError)
  },
)

const apiClient = {
  async delete<T = any>(url: string, config?: ApiRequestConfig): Promise<T> {
    const response = await axiosInstance.delete<ApiResponse<T>>(url, config)
    return response.data.data
  },

  async get<T = any>(url: string, config?: ApiRequestConfig): Promise<T> {
    const response = await axiosInstance.get<ApiResponse<T>>(url, config)
    return response.data.data
  },

  async patch<T = any>(url: string, data?: any, config?: ApiRequestConfig): Promise<T> {
    const response = await axiosInstance.patch<ApiResponse<T>>(url, data, config)
    return response.data.data
  },

  async post<T = any>(url: string, data?: any, config?: ApiRequestConfig): Promise<T> {
    const response = await axiosInstance.post<ApiResponse<T>>(url, data, config)
    return response.data.data
  },

  async put<T = any>(url: string, data?: any, config?: ApiRequestConfig): Promise<T> {
    const response = await axiosInstance.put<ApiResponse<T>>(url, data, config)
    return response.data.data
  },
}

export default apiClient
