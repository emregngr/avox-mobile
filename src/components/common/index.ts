export * from './FaqItem'
export * from './FormElements'
export * from './FullScreenLoading'
export * from './Header'
export * from './OnboardingItem'
export * from './ProfileItem'
export * from './RenderDetailTabBar'
export * from './RenderTabBar'
export * from './SafeLayout'
export * from './SearchInput'
export * from './ThemedButton'
export * from './ThemedButtonText'
export * from './ThemedGradientButton'
export * from './ThemedText'
export * from './ToastConfig'
