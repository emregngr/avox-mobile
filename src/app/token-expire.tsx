import { useQueryClient } from '@tanstack/react-query'
import { Image } from 'expo-image'
import React, { useEffect } from 'react'
import { MMKV } from 'react-native-mmkv'

import { ENUMS } from '@/enums'
import { setIsAuthenticated } from '@/store/auth'
import { responsive } from '@/utils/common/responsive'

const storage = new MMKV()

const Icon = require('@/assets/images/icon-ios.png')

const STATIC_STYLES = {
  icon: {
    height: responsive.deviceHeight,
    width: responsive.deviceWidth,
  },
}

export default function TokenExpire() {
  const queryClient = useQueryClient()

  useEffect(() => {
    const goToAuth = async () => {
      storage.delete(ENUMS.API_TOKEN)
      setIsAuthenticated(false)
      queryClient.setQueryData(['user'], null)
      queryClient.removeQueries()
    }
    goToAuth()
  }, [])

  return (
    <Image
      cachePolicy="memory-disk"
      contentFit="contain"
      source={Icon}
      style={STATIC_STYLES.icon}
      testID="token-expire-icon"
      transition={0}
    />
  )
}
