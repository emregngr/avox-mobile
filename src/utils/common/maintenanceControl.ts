import { Logger } from '@/utils/common/logger'
import type { defaultConfigs } from '@/utils/common/remoteConfig'
import { getBooleanValue, setFirebaseConfig } from '@/utils/common/remoteConfig'

type ConfigKey = keyof typeof defaultConfigs

const getRemoteMaintenance = (): boolean => getBooleanValue('IS_MAINTENANCE' as ConfigKey) ?? false

export const maintenanceControl = async (): Promise<boolean> => {
  try {
    await setFirebaseConfig()
    return getRemoteMaintenance()
  } catch (error) {
    Logger.breadcrumb('maintenanceControlError', 'error', error as Error)
    return false
  }
}
