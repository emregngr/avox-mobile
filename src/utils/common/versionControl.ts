import compareVersions from 'compare-versions'

import Device from '@/utils/common/device'
import { Logger } from '@/utils/common/logger'
import { getStringValue, setFirebaseConfig } from '@/utils/common/remoteConfig'

const isValidVersion = (version: string): boolean => /^\d+\.\d+\.\d+$/.test(version)

const getRemoteVersion = (): string => getStringValue('MIN_VERSION_SUPPORT') ?? '0.0.0'

export const versionControl = async (): Promise<boolean> => {
  try {
    await setFirebaseConfig()
    const minSupportVersion = getRemoteVersion()
    const currentVersion = Device.getVersion()

    if (!isValidVersion(minSupportVersion) || !isValidVersion(currentVersion)) {
      Logger.breadcrumb('Invalid version format', 'info', {
        currentVersion,
        minSupportVersion,
      })
      return false
    }

    const isUpdateAvailable = compareVersions.compare(currentVersion, minSupportVersion, '<')
    return isUpdateAvailable
  } catch (error) {
    Logger.breadcrumb('versionControlError', 'error', error as Error)
    return false
  }
}
