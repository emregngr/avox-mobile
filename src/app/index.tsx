import React from 'react'

import { FullScreenLoading } from '@/components/common'

export default function Index() {
  return <FullScreenLoading />
}
