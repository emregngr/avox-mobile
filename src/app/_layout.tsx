import 'dayjs/locale/en'
import 'dayjs/locale/tr'
import 'react-native-reanimated'
import '../global.css'

import { useReactQueryDevTools } from '@dev-plugins/react-query'
import { ActionSheetProvider } from '@expo/react-native-action-sheet'
import AsyncStorage from '@react-native-async-storage/async-storage'
import crashlytics from '@react-native-firebase/crashlytics'
import * as Sentry from '@sentry/react-native'
import { QueryClientProvider } from '@tanstack/react-query'
import dayjs from 'dayjs'
import customParseFormat from 'dayjs/plugin/customParseFormat'
import localizedFormat from 'dayjs/plugin/localizedFormat'
import timezone from 'dayjs/plugin/timezone'
import utc from 'dayjs/plugin/utc'
import { useFonts } from 'expo-font'
import { router, useFocusEffect } from 'expo-router'
import * as SplashScreen from 'expo-splash-screen'
import { StatusBar } from 'expo-status-bar'
import React, { useCallback, useEffect, useMemo } from 'react'
import { InteractionManager, View } from 'react-native'
import { GestureHandlerRootView } from 'react-native-gesture-handler'
import { KeyboardProvider } from 'react-native-keyboard-controller'
import { SafeAreaProvider } from 'react-native-safe-area-context'
import Toast from 'react-native-toast-message'

import { useToastConfig } from '@/components/common'
import { AppNavigator } from '@/components/feature'
import { queryClient } from '@/config/app/queryClient'
import { ENUMS } from '@/enums'
import { useAppSetup } from '@/hooks/app/useAppSetup'
import { useAppUpdate } from '@/hooks/app/useAppUpdate'
import { useConnectionAlert } from '@/hooks/network/useConnectionAlert'
import { useNotificationSetup } from '@/hooks/notifications/useNotificationSetup'
import { useSystemUI } from '@/hooks/ui/useSystemUI'
import useLocaleStore from '@/store/locale'
import useThemeStore from '@/store/theme'
import { themeColors, themes } from '@/themes'
import { versionControl } from '@/utils/common/useCompareVersion'
import { maintenanceControl } from '@/utils/common/useIsMaintenance'

Sentry.init({
  dsn: process.env.EXPO_PUBLIC_DSN || '',
  enabled: !__DEV__,
})

dayjs.extend(customParseFormat)
dayjs.extend(localizedFormat)
dayjs.extend(utc)
dayjs.extend(timezone)

SplashScreen.preventAutoHideAsync()

const RootLayout = () => {
  const { selectedTheme } = useThemeStore()
  const { selectedLocale } = useLocaleStore()
  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  useReactQueryDevTools(queryClient)

  const [fontsLoaded] = useFonts({
    'Inter-Bold': require('@/assets/fonts/Inter-Bold.ttf'),
    'Inter-Medium': require('@/assets/fonts/Inter-Medium.ttf'),
    'Inter-Regular': require('@/assets/fonts/Inter-Regular.ttf'),
    'Inter-SemiBold': require('@/assets/fonts/Inter-SemiBold.ttf'),
  })

  const { appState, setAppState } = useAppSetup(selectedLocale)
  const toastConfig = useToastConfig(colors)

  useSystemUI(colors, selectedTheme)
  useConnectionAlert({
    isConnected: appState.isConnected,
    onConnectionChange: connected => {
      setAppState(prev => ({
        ...prev,
        isConnected: connected,
      }))
    },
  })
  useAppUpdate()
  useNotificationSetup()

  useFocusEffect(
    useCallback(() => {
      let isActive = true

      const checkAppState = async () => {
        try {
          const [maintenance, versionInvalid] = await Promise.all([
            maintenanceControl(),
            versionControl(),
          ])
          const seenOnboarding = await AsyncStorage.getItem(ENUMS.IS_ONBOARDING_SEEN)

          if (!isActive) return

          if (maintenance) return router.replace('/maintenance')
          if (versionInvalid) return router.replace('/force-update')
          if (!seenOnboarding) return router.replace('/onboarding')

          router.replace('/home')
        } catch (error) {
          if (!__DEV__) crashlytics().recordError(error as Error)
        } finally {
          setAppState(prev => ({ ...prev, isReady: true }))
        }
      }

      InteractionManager.runAfterInteractions(checkAppState)

      return () => {
        isActive = false
      }
    }, [setAppState]),
  )

  const isAppReady =
    fontsLoaded && appState.isConnected && appState.isReady && appState.splashDelayComplete

  useEffect(() => {
    if (isAppReady) {
      InteractionManager.runAfterInteractions(() => {
        SplashScreen.hideAsync()
      })
    }
  }, [isAppReady])

  return (
    <QueryClientProvider client={queryClient}>
      <ActionSheetProvider>
        <SafeAreaProvider>
          <StatusBar
            backgroundColor={colors?.background?.primary}
            style={selectedTheme === 'dark' ? 'light' : 'dark'}
            translucent={false}
          />
          <GestureHandlerRootView className="flex-1">
            <KeyboardProvider>
              <View className="flex-1" style={themes?.[selectedTheme]}>
                <AppNavigator backgroundColor={colors?.background?.blur} isAppReady={isAppReady} />
                <Toast config={toastConfig} />
              </View>
            </KeyboardProvider>
          </GestureHandlerRootView>
        </SafeAreaProvider>
      </ActionSheetProvider>
    </QueryClientProvider>
  )
}

export default Sentry.wrap(RootLayout)
