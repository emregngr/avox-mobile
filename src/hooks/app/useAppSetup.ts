import AsyncStorage from '@react-native-async-storage/async-storage'
import NetInfo from '@react-native-community/netinfo'
import { getApp } from '@react-native-firebase/app'
import { getAuth, signOut } from '@react-native-firebase/auth'
import { getCrashlytics, recordError } from '@react-native-firebase/crashlytics'
import { getMessaging, requestPermission } from '@react-native-firebase/messaging'
import dayjs from 'dayjs'
import {
  getTrackingPermissionsAsync,
  requestTrackingPermissionsAsync,
} from 'expo-tracking-transparency'
import { useCallback, useEffect, useState } from 'react'
import {
  Appearance,
  AppState,
  InteractionManager,
  PermissionsAndroid,
  Platform,
} from 'react-native'
import mobileAds from 'react-native-google-mobile-ads'
import { Notifications } from 'react-native-notifications'

import { ENUMS } from '@/enums'
import { useUserSession } from '@/hooks/app/useUserSession'
import { i18nChangeLocale } from '@/locales/i18next'
import { setIsAuthenticated } from '@/store/auth'
import useLocaleStore from '@/store/locale'
import useThemeStore from '@/store/theme'

interface AppStateType {
  isConnected: boolean | null
  isReady: boolean
  splashDelayComplete: boolean
}

const BADGE_CLEAR_DELAY = {
  BACKGROUND_TO_FOREGROUND: 500,
  INITIAL: 2000,
} as const

const app = getApp()
const messaging = getMessaging(app)
const auth = getAuth()
const crashlytics = getCrashlytics()

export const useAppSetup = () => {
  const { selectedLocale } = useLocaleStore()
  const { selectedTheme } = useThemeStore()

  const [appState, setAppState] = useState<AppStateType>({
    isConnected: true,
    isReady: false,
    splashDelayComplete: false,
  })

  const clearBadge = useCallback(() => {
    try {
      if (Platform.OS === 'ios') {
        Notifications.ios.setBadgeCount(0)
      } else if (Platform.OS === 'android') {
        Notifications.removeAllDeliveredNotifications()
      }
    } catch (error) {}
  }, [])

  const setupIOSTrackingListener = useCallback(async () => {
    if (Platform.OS !== 'ios') return

    const { status } = await getTrackingPermissionsAsync()
    if (status === 'undetermined') {
      const handleTrackingPermission = async (nextAppState: string) => {
        if (nextAppState === 'active') {
          await requestTrackingPermissionsAsync()
          appStateSub?.remove()
        }
      }

      const appStateSub = AppState?.addEventListener('change', handleTrackingPermission)
    }
  }, [])

  const setupNotificationPermissions = useCallback(async () => {
    try {
      await requestPermission(messaging)

      if (Platform.OS === 'android') {
        await PermissionsAndroid.request(PermissionsAndroid.PERMISSIONS.POST_NOTIFICATIONS)
      }
    } catch (error) {
      console.error('Error requesting notification permissions:', error)
    }
  }, [])

  const setupAuthentication = useCallback(async () => {
    try {
      const [token, hasValidSession] = await Promise.all([
        AsyncStorage.getItem(ENUMS.API_TOKEN),
        useUserSession(),
      ])

      if (token && hasValidSession) {
        setIsAuthenticated(true)
      } else {
        await signOut(auth)
        setIsAuthenticated(false)
      }
    } catch (error) {
      await signOut(auth)
      setIsAuthenticated(false)
    }
  }, [])

  const setupApp = useCallback(async () => {
    try {
      Appearance.setColorScheme(selectedTheme)
      dayjs.locale(selectedLocale)
      await i18nChangeLocale(selectedLocale)
      await mobileAds().initialize()

      await Promise.all([
        setupIOSTrackingListener(),
        setupNotificationPermissions(),
        setupAuthentication(),
      ])
    } catch (error) {
      setIsAuthenticated(false)
      if (!__DEV__) {
        recordError(crashlytics, error as Error)
      }
    }
  }, [
    selectedLocale,
    selectedTheme,
    setupIOSTrackingListener,
    setupNotificationPermissions,
    setupAuthentication,
  ])

  const setupNetworkListener = useCallback(
    () =>
      NetInfo?.addEventListener(state => {
        setAppState(prev => ({ ...prev, isConnected: state?.isConnected }))
      }),
    [],
  )

  const setupAppStateListener = useCallback(() => {
    const handleAppStateChange = (nextAppState: string) => {
      if (nextAppState === 'active') {
        setTimeout(clearBadge, BADGE_CLEAR_DELAY.BACKGROUND_TO_FOREGROUND)
      }
    }

    return AppState?.addEventListener('change', handleAppStateChange)
  }, [clearBadge])

  useEffect(() => {
    const networkUnsubscribe = setupNetworkListener()
    const appStateUnsubscribe = setupAppStateListener()

    InteractionManager.runAfterInteractions(() => {
      setupApp()
    })

    return () => {
      networkUnsubscribe?.()
      appStateUnsubscribe?.remove()
    }
  }, [setupApp, setupNetworkListener, setupAppStateListener])

  useEffect(() => {
    const badgeTimer = setTimeout(clearBadge, BADGE_CLEAR_DELAY.INITIAL)

    return () => clearTimeout(badgeTimer)
  }, [clearBadge])

  useEffect(() => {
    const splashTimer = setTimeout(() => {
      setAppState(prev => ({ ...prev, splashDelayComplete: true }))
    }, BADGE_CLEAR_DELAY.INITIAL)

    return () => clearTimeout(splashTimer)
  }, [])

  return { appState, setAppState }
}
