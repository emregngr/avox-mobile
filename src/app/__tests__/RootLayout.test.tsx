import { render, waitFor } from '@testing-library/react-native'
import { useFonts } from 'expo-font'
import * as Linking from 'expo-linking'
import { router } from 'expo-router'
import * as SplashScreen from 'expo-splash-screen'
import { MMKV } from 'react-native-mmkv'
import { Notifications } from 'react-native-notifications'

import RootLayout from '@/app/_layout'
import { useAppSetup } from '@/hooks/app/useAppSetup'
import useThemeStore from '@/store/theme'
import { maintenanceControl } from '@/utils/common/maintenanceControl'
import { versionControl } from '@/utils/common/versionControl'

jest.mock('@rozenite/mmkv-plugin', () => ({
  useMMKVDevTools: jest.fn(),
}))
jest.mock('@rozenite/network-activity-plugin', () => ({
  useNetworkActivityDevTools: jest.fn(),
}))
jest.mock('@rozenite/performance-monitor-plugin', () => ({
  usePerformanceMonitorDevTools: jest.fn(),
}))
jest.mock('@rozenite/tanstack-query-plugin', () => ({
  useTanStackQueryDevTools: jest.fn(),
}))

jest.mock('@/hooks/app/useAppSetup')

jest.mock('@/utils/common/maintenanceControl')

jest.mock('@/utils/common/versionControl')

jest.mock('@/components/feature', () => {
  const { View } = require('react-native')

  return {
    AppNavigator: ({ isAppReady }: { isAppReady: boolean }) =>
      isAppReady ? <View testID="app-navigator" /> : null,
  }
})

jest.mock('@/store/theme')

const mockedUseThemeStore = useThemeStore as jest.MockedFunction<typeof useThemeStore>

const mockedUseAppSetup = useAppSetup as jest.MockedFunction<typeof useAppSetup>
const mockedUseFonts = useFonts as jest.MockedFunction<typeof useFonts>
const mockedMaintenanceControl = maintenanceControl as jest.MockedFunction<
  typeof maintenanceControl
>
const mockedVersionControl = versionControl as jest.MockedFunction<typeof versionControl>

const storage = new MMKV()

const mockedStorageGetString = storage.getString as jest.MockedFunction<typeof storage.getString>

const mockedGetInitialURL = Linking.getInitialURL as jest.MockedFunction<
  typeof Linking.getInitialURL
>
const mockedLinkingParse = Linking.parse as jest.MockedFunction<typeof Linking.parse>
const mockedGetInitialNotification = Notifications.getInitialNotification as jest.MockedFunction<
  typeof Notifications.getInitialNotification
>
const mockedHideAsync = SplashScreen.hideAsync as jest.MockedFunction<typeof SplashScreen.hideAsync>

beforeEach(() => {
  mockedUseThemeStore.mockReturnValue({ selectedTheme: 'light' })

  mockedUseAppSetup.mockReturnValue({
    isConnected: true,
    setIsConnected: jest.fn(),
  })

  mockedUseFonts.mockReturnValue([true] as any)

  mockedMaintenanceControl.mockResolvedValue(false)
  mockedVersionControl.mockResolvedValue(false)

  mockedStorageGetString.mockReturnValue('true')

  mockedGetInitialURL.mockResolvedValue(null)
  mockedGetInitialNotification.mockResolvedValue(null as any)
})

describe('RootLayout', () => {
  it('should render AppNavigator and hide splash screen on successful startup', async () => {
    const { getByTestId } = render(<RootLayout />)

    await waitFor(() => {
      expect(getByTestId('app-navigator')).toBeTruthy()
      expect(mockedHideAsync).toHaveBeenCalled()
      expect(router.replace).toHaveBeenCalledWith('/home')
    })
  })

  it('should route to /maintenance if maintenance mode is active', async () => {
    mockedMaintenanceControl.mockResolvedValue(true)
    render(<RootLayout />)

    await waitFor(() => {
      expect(router.replace).toHaveBeenCalledWith('/maintenance')
    })
  })

  it('should route to /force-update if version is invalid', async () => {
    mockedVersionControl.mockResolvedValue(true)
    render(<RootLayout />)

    await waitFor(() => {
      expect(router.replace).toHaveBeenCalledWith('/force-update')
    })
  })

  it('should route to /onboarding if onboarding has not been seen', async () => {
    mockedStorageGetString.mockReturnValue(null as any)
    render(<RootLayout />)

    await waitFor(() => {
      expect(router.replace).toHaveBeenCalledWith('/onboarding')
    })
  })

  it('should handle deep link to airline-detail', async () => {
    mockedGetInitialURL.mockResolvedValue('yourapp://airline?id=1')

    mockedLinkingParse.mockReturnValue({
      hostname: 'airline',
      queryParams: { id: '1' },
    } as any)

    render(<RootLayout />)

    await waitFor(() => {
      expect(mockedLinkingParse).toHaveBeenCalledWith('yourapp://airline?id=1')
      expect(router.replace).toHaveBeenCalledWith({
        params: { airlineId: '1' },
        pathname: '/airline-detail',
      })
    })
  })

  it('should handle deep link to airport-detail', async () => {
    mockedGetInitialURL.mockResolvedValue('yourapp://airport?id=2')

    mockedLinkingParse.mockReturnValue({
      hostname: 'airport',
      queryParams: { id: '2' },
    } as any)

    render(<RootLayout />)

    await waitFor(() => {
      expect(mockedLinkingParse).toHaveBeenCalledWith('yourapp://airport?id=2')
      expect(router.replace).toHaveBeenCalledWith({
        params: { airportId: '2' },
        pathname: '/airport-detail',
      })
    })
  })

  it('should handle deep link parsing error gracefully and go home in dev mode', async () => {
    mockedGetInitialURL.mockResolvedValue('exp://invalid-url-expo-development-client')
    mockedLinkingParse.mockImplementation(() => {
      throw new Error('Invalid URL')
    })

    render(<RootLayout />)

    await waitFor(() => {
      expect(mockedLinkingParse).toHaveBeenCalledWith('exp://invalid-url-expo-development-client')
      expect(router.replace).toHaveBeenCalledWith('/home')
    })
  })

  it('should handle initial notification for an airline', async () => {
    mockedGetInitialNotification.mockResolvedValue({
      payload: { type: 'airline', id: '1' },
    } as any)
    render(<RootLayout />)

    await waitFor(() => {
      expect(router.replace).toHaveBeenCalledWith({
        params: { airlineId: '1' },
        pathname: '/airline-detail',
      })
    })
  })

  it('should handle initial notification for an airport', async () => {
    mockedGetInitialNotification.mockResolvedValue({
      payload: { type: 'airport', id: '1' },
    } as any)
    render(<RootLayout />)

    await waitFor(() => {
      expect(router.replace).toHaveBeenCalledWith({
        params: { airportId: '1' },
        pathname: '/airport-detail',
      })
    })
  })

  it('should handle expo development client URL correctly', async () => {
    mockedGetInitialURL.mockResolvedValue('exp://192.168.1.1:8081?expo-development-client')

    render(<RootLayout />)

    await waitFor(() => {
      expect(router.replace).toHaveBeenCalledWith('/home')
    })
  })
})

describe('RootLayout Snapshot', () => {
  it('should render the RootLayout successfully', () => {
    const { toJSON } = render(<RootLayout />)

    expect(toJSON()).toMatchSnapshot()
  })
})
