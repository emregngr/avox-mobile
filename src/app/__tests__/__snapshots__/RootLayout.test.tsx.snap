// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`RootLayout Snapshot should render the RootLayout successfully 1`] = `
<View
  className="flex-1"
>
  <View
    className="flex-1"
    style={{}}
  >
    <View
      testID="app-navigator"
    />
  </View>
</View>
`;
